### PR TITLE
controllers delivery service & implement customer gateway & edit listeners accordingly

### LINKED ISSUE #23: Make controllers

### PR DESCRIPTION
**Pull request by arjennorbart**
## Changes
DO NOT MERGE!
This is a draft. Waiting for order and customer service to provide the necessary events/enpoints.

Feel free to check  it out though!

## Checklist
- [x] Implement delivery controller
- [x] Test controllers with Postman
- [x] Add Customer gateway (HttpCustomerRepository) for getting a delivery address from the customer service by REST
- [x] Change Listener to listen to the order queue (Needs to be implemented by the order service, no events are published to the queue at the moment).
- [x] Make command that is used by Listener and will request the delivery address form customer service (need correct endpoint in customer service for this to work)
- [x] Test Listener with rabbitMQ UI (publish message with order.created eventKey and random UUID as orderId)
- [ ] Test everything once required events/endpoints are provided

## When merging, this pull request 
closes #23

### CODE REVIEW

**Review by JustMilan — CHANGES_REQUESTED** [review]
Please rename PR to be more descriptive

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/core/application/DeliveryCommandHandler.java [hunk 1]
```diff
@@ -26,19 +33,27 @@ public void handleStatusChange(UUID id, Status status) {
...
+    // Need to discuss with the boys how to handle fetches appropriately.
+    public void handleRegisterDelivery(UUID orderId) {
+        var customerDeliveryAddress = this.customerGateway.getCustomerDeliveryAddress(orderId);
+
+        var address = new DeliveryAddress(
+                customerDeliveryAddress.street(),
+                customerDeliveryAddress.houseNumber(),
+                customerDeliveryAddress.addition(),
+                customerDeliveryAddress.postalCode(),
+                customerDeliveryAddress.city());
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/core/application/DeliveryCommandHandler.java:52]: nit:
```suggestion
        var address = new DeliveryAddress(
                customerDeliveryAddress.street(),
                customerDeliveryAddress.houseNumber(),
                customerDeliveryAddress.addition(),
                customerDeliveryAddress.postalCode(),
                customerDeliveryAddress.city()
         );
```

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/core/application/DeliveryCommandHandler.java [hunk 2]
```diff
@@ -26,19 +33,27 @@ public void handleStatusChange(UUID id, Status status) {
         this.publishEventsFor(delivery);
     }
 
-    public void handleChangeDeliveryAddress(UUID id, DeliveryAddress address) {
+    public void handleDeliveryAddressChange(UUID id, DeliveryAddress address) {
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/core/application/DeliveryCommandHandler.java:36]: 👍

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/core/port/storage/CustomerRepository.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package nl.softwarestrijders.waiter.delivery.core.port.storage;
+
+import nl.softwarestrijders.waiter.delivery.infrastructure.driven.storage.CustomerDeliveryAddress;
+
+import java.util.UUID;
+
+public interface CustomerRepository {
+    CustomerDeliveryAddress getCustomerDeliveryAddress(UUID orderId);
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/core/port/storage/CustomerRepository.java:8]: Shouldn't it be `deliveryId` or `customerId`? or am i missing something here

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/config/HttpClientConfig.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import nl.softwarestrijders.waiter.delivery.infrastructure.driven.storage.HttpCustomerRepository;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.client.RestTemplate;
+
+@Configuration
+public class HttpClientConfig {
+    @Value("${http-client.root-path.customer}")
+    private String rootPath;
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/config/HttpClientConfig.java:12]: 👍

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/config/RabbitMqConfig.java [hunk 5]
```diff
@@ -24,9 +24,13 @@
 
     @Value("${messaging.queue.delivery}")
     private String deliveryQueueName;
+    @Value("order")
+    private String orderQueueName;
 
     @Value("${messaging.routing-key.delivery}")
     private String deliveryRoutingKey;
+    @Value("order.#")
+    private String orderRoutingKey;
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/config/RabbitMqConfig.java:33]: Could also make properties for these

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driven/storage/CustomerDeliveryAddress.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+import java.util.UUID;
+
+public record CustomerDeliveryAddress(
+        UUID id,
+        int houseNumber,
+        String addition,
+        String street,
+        String postalCode,
+        String city) {
+}
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driven/storage/CustomerDeliveryAddress.java:12]: nit:
```suggestion
public record CustomerDeliveryAddress(
        UUID id,
        int houseNumber,
        String addition,
        String street,
        String postalCode,
        String city
) {
}
```

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driven/storage/HttpCustomerRepository.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+public class HttpCustomerRepository implements CustomerRepository {
+    private final String rootPath;
+    private final RestTemplate client;
+
+    public HttpCustomerRepository(String rootPath, RestTemplate client) {
+        this.rootPath = rootPath;
+        this.client = client;
+    }
+
+    //Need to discuss with Milan what call to make to get the Customer and eventually the CustomerDeliveryAddress.
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driven/storage/HttpCustomerRepository.java:19]: nit:
```suggestion
    // TODO: Need to discuss with Milan what call to make to get the Customer and eventually the CustomerDeliveryAddress.
```

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driven/storage/HttpCustomerRepository.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public CustomerDeliveryAddress getCustomerDeliveryAddress(UUID orderId) {
+        var uri = URI.create(this.rootPath + "/customer/order?order=" + orderId);
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driven/storage/HttpCustomerRepository.java:22]: maybe
```suggestion
        var uri = URI.create(String.format("%s/customer/order?order=%s", this.roothPath, orderId);
```

also, shouldn't it be `rootpath/customer/order/uuid`?

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driver/messaging/RabbitMqEventListener.java [hunk 9]
```diff
@@ -14,11 +14,10 @@ public RabbitMqEventListener(DeliveryCommandHandler commandHandler) {
         this.commandHandler = commandHandler;
     }
 
-    @RabbitListener(queues = "#{'${messaging.queue.delivery}'}")
+    @RabbitListener(queues = "#{'${messaging.queue.order}'}")
     void listen(DeliveryStatusEvent event) {
-        switch (event.eventKey) {
-            case "delivery.status" -> this.commandHandler.handleStatusChange(event.delivery, event.status);
-            case "delivery.address" -> this.commandHandler.handleChangeDeliveryAddress(event.delivery, event.address);
+        if ("order.created".equals(event.eventKey)) {
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driver/messaging/RabbitMqEventListener.java:19]: nit: i would reverse these

File: delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driver/web/DeliveryController.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    public List<DeliveryDto> listAll() {
+        return this.toDto(this.queryHandler.handleGetAllDeliveries());
+    }
+
+    @GetMapping("/{id}")
+    public DeliveryDto findDeliveryById(@PathVariable UUID id) {
+        return this.toDto(this.queryHandler.handleGetDeliveryById(id));
+    }
+
+    @GetMapping("/status/{id}")
```
Comment (xandervedder) [inline delivery/src/main/java/nl/softwarestrijders/waiter/delivery/infrastructure/driver/web/DeliveryController.java:37]:
```suggestion
    @GetMapping("{id}/retrieve-status")
```
you can check the slides about combining REST with CQRS

**Review by xandervedder — CHANGES_REQUESTED** [review]

**Review by xandervedder — APPROVED** [review]

**Review by JustMilan — APPROVED** [review]

**Pull request merged**